pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -335,10 +365,40 @@ func (m *ManagerImpl) Stop() error {
 	return nil
 }
 
-func (m *ManagerImpl) addEndpoint(r *pluginapi.RegisterRequest) {
+func (m *ManagerImpl) addEndpointProbeMode(resourceName string, socketPath string) (chan bool, error) {
+	chanForAckOfNotification := make(chan bool)
+
+	old, new, err := m.createEndpoint(resourceName, socketPath)
+	if err != nil {
+		glog.Errorf("Failed to dial device plugin with socketPath %s: %v", socketPath, err)
+		return nil, fmt.Errorf("Failed to dial device plugin with socketPath %s: %v", socketPath, err)
+	}
+
+	options, err := new.client.GetDevicePluginOptions(context.Background(), &pluginapi.Empty{})
+	if err != nil {
+		glog.Errorf("Failed to get device plugin options: %v", err)
+		return nil, fmt.Errorf("Failed to get device plugin options: %v", err)
+	}
+	if err := m.registerEndpoint(resourceName, options, old, new); err != nil {
+		return nil, err
+	}
+
+	go func() {
+		select {
+		case <-chanForAckOfNotification:
+			close(chanForAckOfNotification)
+			m.runEndpoint(resourceName, old, new)
+		case <-time.After(time.Second):
+			glog.Errorf("Timed out while waiting for notification ack from plugin")
+		}
+	}()
+	return chanForAckOfNotification, nil
+}
+
+func (m *ManagerImpl) createEndpoint(resourceName string, socketPath string) (endpoint, *endpointImpl, error) {
 	existingDevs := make(map[string]pluginapi.Device)
```

Comment:
As already mentioned, the general code path should be:

```golang
e := newEndpointImpl(...)
options := e.GetDevicePluginOptions(...)

m.mutex.Lock()
defer m.mutex.Unlock()

var devs map[string]pluginapi.Device
old, ok := m.endpoints[resourceName]
if ok {
    devs = old.Devices()
}

m.endpoints[resourceName] = e

if ok {
    old.Stop()
}
```

This makes the whole re-registration issue caused because you are locking twice go away.

Reply:
had to reorganize refactoring. should be good now. PTAL!